source/common/common/inline_map_registry.h

```diff
@@ -0,0 +1,164 @@
+#pragma once
+
+#include "source/common/common/inline_map.h"
+
+namespace Envoy {
+
+/**
+ * Registry class to get inline map descriptor singleton based on the scope. If cross-modules inline
+ * map descriptor is necessary, this class should be used.
+ *
+ * There are only two allowed ways to update or modify the registry:
+ * 1. Add inline map descriptors or inline keys to the registry before the main() function is called
+ *    by the REGISTER_INLINE_MAP_DESCRIPTOR or REGISTER_INLINE_MAP_KEY macro.
+ * 2. Add inline map descriptors or inline keys in the finalize() function.
+ *
+ * The finalize() function will be called once when the first server is initialized. This is safe
+ * even there are multiple main threads or servers.
+ */
+class InlineMapRegistry {
+public:
+  /**
+   * Helper template class to initialize descriptor based on the scope explicitly.
+   */
+  template <class Scope> class InlineMapDescriptorRegister {
+  public:
+    InlineMapDescriptorRegister();
+  };
+
+  /**
+   * Helper template class to add inline key to the descriptor based on the scope.
+   */
+  template <class Scope> class InlineMapKeyRegister {
+  public:
+    InlineMapKeyRegister(absl::string_view);
+  };
+
+  using InlineMapDescriptor = InlineMapDescriptor<std::string>;
+  using Handle = InlineMapDescriptor::Handle;
+
+  struct ScopeInlineKeys {
+    std::string name;
+    absl::flat_hash_set<std::string> keys;
+  };
+
+  /**
+   * Scope inline keys vector type. This is used to finalize the registry and add inline keys to
+   * the descriptor based on the scope name.
+   */
+  using ScopeInlineKeysVector = std::vector<ScopeInlineKeys>;
+
+  /**
+   * Registry info type. This is used to get scope name and all managed descriptors' inline keys as
+   * string.
+   */
+  using RegistryInfo = std::vector<std::pair<std::string, std::string>>;
+
+private:
+  using Registry = absl::flat_hash_map<std::string, InlineMapDescriptor*>;
+
+  template <class Scope> friend class InlineMapDescriptorRegister;
+  template <class Scope> friend class InlineMapKeyRegister;
+
+  /**
+   * Call once flag to ensure the registry is finalized only once on multi-thread environment.
+   */
+  static std::once_flag& onceFlag() { MUTABLE_CONSTRUCT_ON_FIRST_USE(std::once_flag); }
```

Comment:
I don't love this pattern of having lots of MUTABLE_CONSTRUCT_ON_FIRST_USE in separate static member functions. How about just one MUTABLE_CONSTRUCT_ON_FIRST_USE for the class instance then use normal accessors on that class for each member var.

Remember each MUTABLE_CONSTRUCT_ON_FIRST_USE needs to do some kind of lock or atomic to avoid races, so it's better to just do one of them.

Reply:
I can have a try.